pm: Add README

yondonfu opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->

This PR adds:

- A brief README for the `pm` package that references the full PM protocol specification and also highlights a few things that the package implements to ensure fair winning ticket selection.
- A payment doc in `doc/payments.md` that includes links to resources for the PM protocol as well as notes on fee estimation

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->

See above.

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

N/A

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

j0sh commented:
The first part looks pretty good to me. The RNG part feels a little ad hoc / incomplete without some more context around PM. Even though there is a link to the full PM spec, it might not be immediately clear to a new reader where the RNG part fits in and the reasoning for certain behaviors.

Eg, from a quick glance, the writeup mentions that winning `recipientRand` values are only persisted in memory. A sharp reader might ask why "in memory" is a sufficient defense, and the reader needs to connect a few dots themselves within the full spec (or via reading the code) to figure out it's because `O.secret` is generated fresh at each start-up time.

Not really sure yet how/where to improve this without a closer read of the PM spec.

kyriediculous commented on pm/README.md:
```diff
@@ -0,0 +1,37 @@
+
+The `abi.encodePacked()` Solidity built-in is used to match the behavior of the contract implementing the `TicketBroker` interface. `senderSig` is a signature produced by `Sender` over the hash of the ticket and `recipientRand` is the pre-image of the `recipientRandHash` that is included in the ticket.
+
+Before a `Sender` can start creating and sending tickets, it must first fetch ticket parameters from a `Recipient`. When generating ticket parameters, a `Recipient` will generate a random number `recipientRand`, compute the hash of `recipientRand` as `recipientRandHash` and include the hash with the ticket parameters. Once a `Sender` has a set of ticket parameters with a `recipientRandHash`, the  ...
```
It might be a good idea to provide some extra context around `TicketParams` here , perhaps we can use the commented protobuf definitions to provide more clarity here?

kyriediculous commented on pm/README.md:
```diff
@@ -0,0 +1,37 @@
+
+The `pm` package contains a client implementation of the [Livepeer PM protocol](https://github.com/livepeer/wiki/blob/master/spec/streamflow/pm.md). The main areas of interest are:
+
+- A `Ticket` type which defines the ticket format that both senders and recipients accept.
```
It might be a good idea to provide some extra context around this type as its fieldnames are used extensively in the RNG part of the README. WDYT?

thomshutt commented:
@yondonfu Is this documentation still worth getting merged or can we close the PR?

yondonfu commented:
Reviving this PR! Updated the `pm` package docs with some tweaks based on recent changes (i.e. https://github.com/livepeer/go-livepeer/pull/1383 and https://github.com/livepeer/go-livepeer/pull/2710) as well as an additional payments doc with links to PM resources and some notes on fee estimation. These docs are by no means comprehensive and can certainly be improved, but think it should be a 0 -> 1 upgrade that will be a good start for now!

thomshutt reviewed (approved):
This is great, thanks Yondon! I think you also win longest time between PR opening and getting merged 🏆

yondonfu committed:
pm: Add README

yondonfu committed:
doc: Add payments doc

yondonfu commented:
Made a few small changes:

- Moved diagram into `pm/assets` folder and included a text file with the sequence diagram in UML so that it can be updated later if needed
- Moved the diagram PNG file into git LFS

The above changes align with the convention used for assets in docs seen in `doc/assets`. They also seem to fix a weird dirty git diff error I was encountering in CI builds...
